### PR TITLE
Create GoogleMusicImporter and GoogleMusicExporter

### PR DESCRIPTION
**Pull request by gauss0818**
Add google music import/export adapters.

Changlist:
1. org/datatransferproject/datatransfer/google/musicModels/.. - Add classes to interpret requests/responses to/from Google Music public API.
2. org/datatransferproject/datatransfer/google/music/.. - Add adapters and an interface to call Google Music public API.
3. unit tests for GoogleMusicImporter and GoogleMusicExporter

### CODE REVIEW

**Review by jzacsh — CHANGES_REQUESTED** [review]

File: extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/music/GoogleMusicExporter.java [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+      for (GooglePlaylistItem googlePlaylistItem : googlePlaylistItems) {
+        playlistItems.add(convertPlaylistItem(playlistId, googlePlaylistItem));
+      }
+      containerResource = new MusicContainerResource(null, playlistItems, null, null);
+    }
+
+    return new ExportResult<>(ResultType.CONTINUE, containerResource, continuationData);
+  }
+
+  private int getTokenPrefixLength(String token) {
```
[the remark this answers is not in the document]
Comment (gauss0818) [reply extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/music/GoogleMusicExporter.java:218]: Done.

File: extensions/data-transfer/portability-data-transfer-google/src/test/java/org/datatransferproject/datatransfer/google/music/GoogleMusicExporterTest.java [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+import org.datatransferproject.types.common.models.music.MusicContainerResource;
+import org.datatransferproject.types.common.models.music.MusicPlaylist;
+import org.datatransferproject.types.common.models.music.MusicPlaylistItem;
+import org.datatransferproject.types.common.models.music.MusicRecording;
+import org.datatransferproject.types.common.models.music.MusicRelease;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+public class GoogleMusicExporterTest {
+  static final String PLAYLIST_TOKEN = "playlist_token";
```
[the remark this answers is not in the document]
Comment (gauss0818) [reply extensions/data-transfer/portability-data-transfer-google/src/test/java/org/datatransferproject/datatransfer/google/music/GoogleMusicExporterTest.java]: You're right. Updated!

File: extensions/data-transfer/portability-data-transfer-google/src/test/java/org/datatransferproject/datatransfer/google/music/GoogleMusicExporterTest.java [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+    // Check pagination token
+    ContinuationData continuationData = result.getContinuationData();
+    PaginationData paginationToken = continuationData.getPaginationData();
+    assertThat(paginationToken).isNull();
+  }
+
+  /** Sets up a response with a single playlist, containing a single playlist item */
+  private void setUpSinglePlaylist() {
+    GooglePlaylist playlistEntry = new GooglePlaylist();
+    playlistEntry.setName(GOOGLE_PLAYLIST_NAME_PREFIX + "p1_id");
```
[the remark this answers is not in the document]
Comment (gauss0818) [reply extensions/data-transfer/portability-data-transfer-google/src/test/java/org/datatransferproject/datatransfer/google/music/GoogleMusicExporterTest.java]: Done.

File: extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/musicModels/GooglePlaylist.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+  public long getCreateTime() {
+    return createTime;
+  }
+
+  public long getUpdateTime() {
+    return updateTime;
+  }
+
+  public String getToken() {
```
[the remark this answers is not in the document]
Comment (gauss0818) [reply extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/musicModels/GooglePlaylist.java:62]: It does not mean the pagination token. The token inside GooglePlaylist is used by the public API to find the playlist easily when inserting the playlist items. I added a TODO to rename the this field. We're still discussing the naming for the token internally.

File: extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/music/GoogleMusicImporter.java [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+      idempotentImportExecutor.executeAndSwallowIOExceptions(
+          playlist.getId(),
+          playlist.getTitle(),
+          () -> importSinglePlaylist(jobId, authData, playlist));
+    }
+
+    // Create playlistItems
+    importPlaylistItems(data.getPlaylistItems(), idempotentImportExecutor, jobId, authData);
+
+    // TODO: create tracks
```
[the remark this answers is not in the document]
Comment (gauss0818) [reply extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/music/GoogleMusicImporter.java:121]: Updated necessary TODOs as TODO(critical WIP-feature step).

File: extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/music/GoogleMusicInterface.java [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+import org.datatransferproject.datatransfer.google.common.GoogleCredentialFactory;
+import org.datatransferproject.datatransfer.google.musicModels.BatchPlaylistItemRequest;
+import org.datatransferproject.datatransfer.google.musicModels.BatchPlaylistItemResponse;
+import org.datatransferproject.datatransfer.google.musicModels.GooglePlaylist;
+import org.datatransferproject.datatransfer.google.musicModels.PlaylistItemListResponse;
+import org.datatransferproject.datatransfer.google.musicModels.PlaylistListResponse;
+import org.datatransferproject.spi.transfer.types.InvalidTokenException;
+import org.datatransferproject.spi.transfer.types.PermissionDeniedException;
+
+public class GoogleMusicInterface {
```
[the remark this answers is not in the document]
Comment (gauss0818) [reply extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/music/GoogleMusicInterface.java]: Updated as GoogleMusicHttpApi.

File: extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/music/GoogleMusicInterface.java [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+  BatchPlaylistItemResponse createPlaylistItems(BatchPlaylistItemRequest playlistItemRequest)
+      throws IOException, InvalidTokenException, PermissionDeniedException {
+    Map<String, Object> playlistItemMap = createJsonMap(playlistItemRequest);
+    HttpContent content = new JsonHttpContent(jsonFactory, playlistItemMap);
+
+    return makePostRequest(
+        BASE_URL + "playlistItems", Optional.empty(), content, BatchPlaylistItemResponse.class);
+  }
+
+  private <T> T makeGetRequest(String url, Optional<Map<String, String>> parameters, Class<T> clazz)
```
[the remark this answers is not in the document]
Comment (gauss0818) [reply extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/music/GoogleMusicInterface.java]: Done

File: extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/music/GoogleTempPlaylistToken.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+package org.datatransferproject.datatransfer.google.music;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonTypeName;
+import org.datatransferproject.types.common.models.DataModel;
+
+/**
+ * Class representing a temporary Google Playlist Token which will be stored in each job's memory.
```
[the remark this answers is not in the document]
Comment (gauss0818) [reply extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/music/GoogleTempPlaylistToken.java:25]: Done

File: extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/musicModels/GoogleRelease.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+ * limitations under the License.
+ */
+
+package org.datatransferproject.datatransfer.google.musicModels;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import java.util.Arrays;
+import java.util.Objects;
+
+/** Class representing a release as returned by the Google Music API. */
```
[the remark this answers is not in the document]
Comment (gauss0818) [reply extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/musicModels/GoogleRelease.java]: Done

File: extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/musicModels/GoogleTrack.java [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+ * limitations under the License.
+ */
+
+package org.datatransferproject.datatransfer.google.musicModels;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import java.util.Arrays;
+import java.util.Objects;
+
+/** Class representing a track as returned by the Google Music API. */
```
[the remark this answers is not in the document]
Comment (gauss0818) [reply extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/musicModels/GoogleTrack.java]: Done

File: extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/musicModels/Status.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+package org.datatransferproject.datatransfer.google.musicModels;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+/** Information about whether a request was successful, used by {@link NewPlaylistItemResult}. */
+public class Status {
+
+  @JsonProperty("code")
+  private int code;
```
[the remark this answers is not in the document]
Comment (gauss0818) [reply extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/musicModels/Status.java:28]: Got it. We will have the documentation for the API after we finalized the details.

File: extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/musicModels/Status.java [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.datatransferproject.datatransfer.google.musicModels;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+/** Information about whether a request was successful, used by {@link NewPlaylistItemResult}. */
```
[the remark this answers is not in the document]
Comment (gauss0818) [reply extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/musicModels/Status.java]: Added a TODO.

File: extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/music/GoogleMusicExporter.java [hunk 13]
```diff
@@ -0,0 +1,281 @@
...
+      return new ExportResult<>(ResultType.END, null, null);
+    }
+  }
+
+  @VisibleForTesting
+  ExportResult<MusicContainerResource> exportPlaylists(
+      TokensAndUrlAuthData authData, Optional<PaginationData> paginationData, UUID jobId)
+      throws IOException, InvalidTokenException, PermissionDeniedException {
+    Optional<String> paginationToken = Optional.empty();
+    String prefixToken = "";
```
[the remark this answers is not in the document]
Comment (gauss0818) [reply extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/music/GoogleMusicExporter.java]: Done

File: extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/music/GoogleMusicExporter.java [hunk 14]
```diff
@@ -0,0 +1,281 @@
...
+    PaginationData nextPageData;
+    String token = playlistListResponse.getNextPageToken();
+    List<MusicPlaylist> playlists = new ArrayList<>();
+    GooglePlaylist[] googlePlaylists = playlistListResponse.getPlaylists();
+
+    if (Strings.isNullOrEmpty(token)) {
+      nextPageData =
+          new StringPaginationToken(prefixToken.substring(PLAYLIST_TOKEN_PREFIX.length()));
+    } else {
+      nextPageData = new StringPaginationToken(prefixToken + token);
```
Comment (jzacsh) [inline extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/music/GoogleMusicExporter.java]: per offline chat; this conditional should have one more line: `ResultType resultType = ResultType.CONTINUE;`

File: extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/music/GoogleMusicExporter.java [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+    PlaylistListResponse playlistListResponse =
+        getOrCreateMusicInterface(authData).listPlaylists(paginationToken);
+
+    PaginationData nextPageData;
+    String token = playlistListResponse.getNextPageToken();
+    List<MusicPlaylist> playlists = new ArrayList<>();
+    GooglePlaylist[] googlePlaylists = playlistListResponse.getPlaylists();
+
+    if (Strings.isNullOrEmpty(token)) {
+      nextPageData =
```
Comment (jzacsh) [inline extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/music/GoogleMusicExporter.java:151]: per offline chat; this conditional should have one more line: `ResultType resultType = ResultType.END;`

File: extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/music/GoogleMusicExporter.java [hunk 16]
```diff
@@ -0,0 +1,281 @@
...
+            () ->
+                String.format(
+                    "%s: Google Music exporting playlist: %s", jobId, musicPlaylist.getId()));
+
+        // Add playlist id to continuation data
+        continuationData.addContainerResource(new IdOnlyContainerResource(musicPlaylist.getId()));
+      }
+    }
+
+    ResultType resultType = ResultType.CONTINUE;
```
Comment (jzacsh) [inline extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/music/GoogleMusicExporter.java:179]: per offline chat we should remove this and define it in the conditional (that defines your `nextPageData`) above

**Review by jzacsh — DISMISSED** [review]
some changes pending but otherwise LGTM!

File: extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/music/GoogleMusicExporter.java [hunk 17]
```diff
@@ -0,0 +1,281 @@
...
+        && paginationToken.getToken().startsWith(RELEASE_TOKEN_PREFIX)) {
+      // TODO: export releases
+      return new ExportResult<>(ResultType.END, null, null);
+    } else {
+      // There is nothing to export.
+      return new ExportResult<>(ResultType.END, null, null);
+    }
+  }
+
+  @VisibleForTesting
```
[the remark this answers is not in the document]
Comment (gauss0818) [reply extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/music/GoogleMusicExporter.java:128]: Added TODO for @jzacsh

File: extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/music/GoogleMusicExporter.java (hunk 14 above)
Comment (gauss0818) [reply extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/music/GoogleMusicExporter.java]: Updated resultType based on the nextPageData condition.

File: extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/music/GoogleMusicExporter.java (hunk 16 above)
Comment (gauss0818) [reply extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/music/GoogleMusicExporter.java]: Updated resultType based on the nextPageData condition.

File: extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/music/GoogleTempPlaylistToken.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (gauss0818) [reply extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/music/GoogleTempPlaylistToken.java]: Renamed it as GooglePlaylistInsertionToken.

File: extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/music/GoogleMusicExporter.java [hunk 18]
```diff
@@ -0,0 +1,281 @@
...
+        getOrCreateMusicInterface(authData).listPlaylists(paginationToken);
+
+    PaginationData nextPageData;
+    String token = playlistListResponse.getNextPageToken();
+    List<MusicPlaylist> playlists = new ArrayList<>();
+    GooglePlaylist[] googlePlaylists = playlistListResponse.getPlaylists();
+
+    if (Strings.isNullOrEmpty(token)) {
+      nextPageData =
+          new StringPaginationToken(prefixToken.substring(PLAYLIST_TOKEN_PREFIX.length()));
```
[the remark this answers is not in the document]
Comment (gauss0818) [reply extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/music/GoogleMusicExporter.java]: If there is no pending playlists to fetch, the prefixToken will remove "playlist:". 
Example:
"playlist:release:" -> "release:" (finish exporting playlists, start exporting releases)

If we uses prefix token to indicate exporting playlist items:
"playlist:playlist_item:" -> "playlist_item:"

**Review by jzacsh — APPROVED** [review]

File: extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/music/GoogleMusicExporter.java [hunk 19]
```diff
@@ -0,0 +1,277 @@
...
+                release.getIcpn(),
+                release.getReleaseTitle(),
+                createMusicGroups(release.getArtistTitles())),
+            createMusicGroups(track.getArtistTitles())),
+        playlistId,
+        googlePlaylistItem.getOrder());
+  }
+
+  private synchronized GoogleMusicHttpApi getOrCreateMusicHttpApi(TokensAndUrlAuthData authData) {
+    return musicHttpApi == null ? makeMusicHttpApi(authData) : musicHttpApi;
```
Comment (jzacsh) [inline extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/music/GoogleMusicExporter.java:269]: note for myself when I merge this later: need to fix `musicHttpApi` as it never gets set

**Pull request merged**